astropy/timeseries/downsample.py

```diff
@@ -0,0 +1,136 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+import warnings
+
+import numpy as np
+from astropy import units as u
+from astropy.utils.exceptions import AstropyUserWarning
+
+from astropy.timeseries.sampled import TimeSeries
+from astropy.timeseries.binned import BinnedTimeSeries
+
+__all__ = ['simple_downsample']
+
+
+def reduceat(array, indices, function):
+    """
+    Manual reduceat functionality for cases where Numpy functions don't have a reduceat.
+    It will check if the input function has a reduceat and call that if it does.
+    """
+    if hasattr(function, 'reduceat'):
+        return np.array(function.reduceat(array, indices))
+    else:
+        result = []
+        for i in range(len(indices) - 1):
+            if indices[i+1] <= indices[i]+1:
+                result.append(function(array[indices[i]]))
+            else:
+                result.append(function(array[indices[i]:indices[i+1]]))
+        result.append(function(array[indices[-1]:]))
+        return np.array(result)
+
+
+def simple_downsample(time_series, time_bin_size, func=None, time_bin_start=None, n_bins=None):
+    """
+    Downsample a time series by binning values into bins with a fixed size,
+    using a single function
+
+    Parameters
+    ----------
+    time_series : :class:`~astropy.timeseries.TimeSeries`
+        The time series to downsample.
+    time_bin_size : `~astropy.units.Quantity`
```

Comment:
I think there should be a way to call this that lets the user pass in a custom binning scheme.  That would mean this argument should instead be `time_bin`, which if not a scalar, would mean "use these bins" (and would raise an exception if `time_bin_start` or `n_bins` are not None.

Otherwise, `func` really ought to be the *second* argument.  It's confusing that func is in the middle of the argument list even though `time_bin_start` and `n_bins` are more related to the bin size than the `func`.

Reply:
I agree with this, however I think we can punt on it for now and add support for custom binning later. I've re-ordered the arguments to have a more natural order and made all the arguments except the first keyword-only to make sure we don't run into API issues in future though.

Reply:
Issue here: https://github.com/astropy/astropy/issues/8584